lib/runtime-core/src/parse.rs

```diff
@@ -110,11 +110,46 @@ pub fn read_module<
     let mut namespace_builder = Some(StringTableBuilder::new());
     let mut name_builder = Some(StringTableBuilder::new());
     let mut func_count: usize = 0;
+    let mut mcg_signatures_fed = false;
     let mut mcg_info_fed = false;
 
     loop {
         use wasmparser::ParserState;
         let state = parser.read();
+
+        // Feed signature and namespace information as early as possible.
+        match *state {
+            ParserState::BeginFunctionBody { .. }
+            | ParserState::ImportSectionEntry { .. }
+            | ParserState::EndWasm => {
+                if !mcg_signatures_fed {
+                    mcg_signatures_fed = true;
+                    let info_read = info.read().unwrap();
+                    mcg.feed_signatures(info_read.signatures.clone())
+                        .map_err(|x| LoadError::Codegen(format!("{:?}", x)))?;
+                }
+            }
+            _ => {}
+        }
+        match *state {
+            ParserState::BeginFunctionBody { .. } | ParserState::EndWasm => {
+                if !mcg_info_fed {
```

Comment:
If we use separate `match` statements, moving the `if !mcg_*_fed {` checks outside the `match *state` will probably have better performance due to better branch prediction.

Reply:
I think this is a kind of simple optimization that can be left to the compiler?

Reply:
Depends whether the compiler can prove `*match` is side-effect free. I'm sorta torn on this one, I think we should write whatever is clearest to the reader, but I'm not sure that pulling this part of the match out in advance is actually clearer. I'm happy to let you and Mark decide.

Reply:
Will make it into a closure.